data_safe_haven/config/context_settings.py

```diff
@@ -139,41 +154,54 @@ def add(
             msg = f"A context with key '{key}' is already defined."
             raise DataSafeHavenParameterError(msg)
 
-        self.settings["contexts"][key] = {
-            "name": name,
-            "admin_group_id": admin_group_id,
-            "location": location,
-            "subscription_name": subscription_name,
-        }
+        self.contexts[key] = Context(
+            name=name,
+            admin_group_id=admin_group_id,
+            location=location,
+            subscription_name=subscription_name,
+        )
 
     def remove(self, key: str) -> None:
         if key not in self.available:
             msg = f"No context with key '{key}'."
             raise DataSafeHavenParameterError(msg)
-        del self.settings["contexts"][key]
+        del self.contexts[key]
```

Comment:
Make sure that if you remove the currently selected context, you delete the selected field from the `yaml` too. If you remove the currently selected context, you then get errors when trying to do anything else with contexts, because the selected context is then undefined: 

```
(data-safe-haven) deploydsh@af988d5aec18:/workspaces/data-safe-haven$ dsh context remove testing
2023-11-29 14:14:58 [    INFO] Reading project settings from '/home/deploydsh/.config/data_safe_haven/contexts.yaml'.           context_settings.py:192
2023-11-29 14:14:58 [    INFO] Saved context settings to '/home/deploydsh/.config/data_safe_haven/contexts.yaml'.               context_settings.py:215
(data-safe-haven) deploydsh@af988d5aec18:/workspaces/data-safe-haven$ dsh context add testing --admin-group 347c68cb-261f-4a3e-ac3e-6af860b5fec9 --location uksouth --name green --subscription "Data Safe Haven Development"
2023-11-29 14:15:03 [    INFO] Reading project settings from '/home/deploydsh/.config/data_safe_haven/contexts.yaml'.           context_settings.py:192
2023-11-29 14:15:03 [   ERROR] Could not load context settings.                                                                               cli.py:92
2023-11-29 14:15:03 [   ERROR] 1 validation error for ContextSettings                                                                         cli.py:92
2023-11-29 14:15:03 [   ERROR]   Value error, Selected context 'testing' is not defined.                                                      cli.py:92
2023-11-29 14:15:03 [   ERROR]     For further information visit https://errors.pydantic.dev/2.5/v/value_error     
```